identifier/identifier_test.go

```diff
@@ -0,0 +1,136 @@
+package identifier
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"net"
+	"net/netip"
+	"testing"
+
+	"github.com/letsencrypt/boulder/test"
+)
+
+// TestFromX509 tests FromCert and FromCSR, which are fromX509's public
+// wrappers.
+func TestFromX509(t *testing.T) {
+	cases := []struct {
+		name           string
+		subject        pkix.Name
+		dnsNames       []string
+		ipAddresses    []net.IP
+		expectedIdents []ACMEIdentifier
+	}{
+		{
+			name:           "no explicit CN",
+			dnsNames:       []string{"a.com"},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com")},
+		},
+		{
+			name:           "explicit uppercase CN",
+			subject:        pkix.Name{CommonName: "A.com"},
+			dnsNames:       []string{"a.com"},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com")},
+		},
+		{
+			name:           "no explicit CN, uppercase SAN",
+			dnsNames:       []string{"A.com"},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com")},
+		},
+		{
+			name:           "duplicate SANs",
+			dnsNames:       []string{"b.com", "b.com", "a.com", "a.com"},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com"), NewDNS("b.com")},
+		},
+		{
+			name:           "explicit CN not found in SANs",
+			subject:        pkix.Name{CommonName: "a.com"},
+			dnsNames:       []string{"b.com"},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com"), NewDNS("b.com")},
+		},
+		{
+			name:           "mix of DNSNames and IPAddresses",
+			dnsNames:       []string{"a.com"},
+			ipAddresses:    []net.IP{{192, 168, 1, 1}},
+			expectedIdents: []ACMEIdentifier{NewDNS("a.com"), NewIP(netip.MustParseAddr("192.168.1.1"))},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			cert := &x509.Certificate{Subject: tc.subject, DNSNames: tc.dnsNames, IPAddresses: tc.ipAddresses}
+			csr := &x509.CertificateRequest{Subject: tc.subject, DNSNames: tc.dnsNames, IPAddresses: tc.ipAddresses}
+			test.AssertDeepEquals(t, FromCert(cert), tc.expectedIdents)
+			test.AssertDeepEquals(t, FromCSR(csr), tc.expectedIdents)
+		})
+	}
```

Comment:
Here's a suggestion to do this more in line with our recent best practices (avoiding test.Assert, using got/want instead of actual/expect) and with some extra parallelism thrown in
```suggestion
	for _, tc := range cases {
		t.Run("cert/"+tc.name, func(t *testing.T) {
			t.Parallel()
			got := FromCert(&x509.Certificate{Subject: tc.subject, DNSNames: tc.dnsNames, IPAddresses: tc.ipAddresses})
			if !slices.Equal(got, want) {
				t.Errorf("FromCert() got %#v, but want %#v", got, tc.want)
			}
		})
		t.Run("csr/"+tc.name, func(t *testing.T) {
			t.Parallel()
			got := FromCSR(&x509.CertificateRequest{Subject: tc.subject, DNSNames: tc.dnsNames, IPAddresses: tc.ipAddresses})
			if !slices.Equal(got, want) {
				t.Errorf("FromCSR() got %#v, but want %#v", got, tc.want)
			}	
		})
	}
```